word-level-pos-tokens-recogs-style-decoder-loop.rasp

```diff
@@ -716,10 +729,52 @@ output = "(" if ((num_tokens_in_output_excluding_asterisks % 5) == 1) else outpu
 output = target_word_index if ((num_tokens_in_output_excluding_asterisks % 5) == 2) else output;
 output = ")" if ((num_tokens_in_output_excluding_asterisks % 5) == 3) else output;
 # note that when nv_in_output_count == nv_in_input_count, we will add AND instead of ";"
-output = (";" if (5 * nv_in_input_count - 1 > num_tokens_in_output_excluding_asterisks) else "AND") if (num_tokens_in_output_excluding_asterisks % 5 == 4) else output;
-
-# if we didn't have an input/output separator that needs to be output
-output = "|" if num_pipes_in_output == 0 else output;
+output = (";" if (5 * noun_in_input_count - 1 >= num_tokens_in_output_excluding_asterisks) else "AND") if (num_tokens_in_output_excluding_asterisks % 5 == 4) else output;
+
+# CP handling
+# (nv_in_output_count > noun_in_input_count) and (nv_in_output_count <=  noun_in_input_count + cp_in_input_count)
+cp_skip_tokens = 5 * noun_in_input_count;
+# mean ( 1 ) AND agent ( 1 , 0 ) AND ccomp (   1  ,   4 ) AND
+#    1 2 3 4   5     6 7 8 9 10 11 12   13 14 15 16  17 18 19
+cp_block_num_tokens = 19; # see above, fact of dataset, not my choice
+
+# CP logical form template (large since completely fixed so no need to do any dynamic lookup for each part)
+cp_output = target_word_token if (((num_tokens_in_output_excluding_asterisks-cp_skip_tokens) % cp_block_num_tokens) == 0) else "";
+cp_output = "(" if (((num_tokens_in_output_excluding_asterisks-cp_skip_tokens) % cp_block_num_tokens) == 1) else cp_output;
+cp_output = target_word_index if (((num_tokens_in_output_excluding_asterisks-cp_skip_tokens) % cp_block_num_tokens) == 2) else cp_output;
+cp_output = ")" if ((num_tokens_in_output_excluding_asterisks-cp_skip_tokens) % cp_block_num_tokens == 3) else cp_output;
+cp_output = "AND" if ((num_tokens_in_output_excluding_asterisks-cp_skip_tokens) % cp_block_num_tokens == 4) else cp_output;
+cp_output = "agent" if ((num_tokens_in_output_excluding_asterisks-cp_skip_tokens) % cp_block_num_tokens == 5) else cp_output;
+cp_output = "(" if ((num_tokens_in_output_excluding_asterisks-cp_skip_tokens) % cp_block_num_tokens == 6) else cp_output;
+cp_agent_left_idx = target_word_index;
+cp_output = cp_agent_left_idx if ((num_tokens_in_output_excluding_asterisks-cp_skip_tokens) % cp_block_num_tokens == 7) else cp_output;
+cp_output = "," if ((num_tokens_in_output_excluding_asterisks-cp_skip_tokens) % cp_block_num_tokens == 8) else cp_output;
+cp_idx = cp_in_output_count;
+cp_target_noun_index = aggregate(select(indices, cp_idx, ==), input_indices_sorted);
+cp_agent_right_idx = cp_target_noun_index; # this is the ith noun where i is the cp index we are on
```

Comment:
Needs to ignore "pp np" when getting relationship index as already done elsewhere, otherwise will be broken by prepositional phrase modification within complement phrase

Reply:
`cp_target_noun_index = aggregate(select(nps_without_pp_prefix_indices, cp_in_output_count+1, ==), indices);`, I think

and move the definition of `nps_without_pp_prefix_indices` above this.

Reply:
I can't push this change up safely due to power outage, unfortunately